Update to Polkadot v0.9.29

NunoAlexandre opened the pull request:
# Description

Upgrading `centrifuge-chain` to the latest Polkadot version, `v0.9.29`

Fixes https://github.com/centrifuge/centrifuge-chain/issues/976

## Changes and Descriptions

This change also includes the upgrade to `v0.9.28`: https://github.com/paritytech/polkadot/releases/tag/v0.9.28

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] `cargo build --release` 
- [x] `cargo test --workspace --release --features test-benchmarks,try-runtime,runtime-benchmarks`

# Checklist:

- [ ] I have added Rust doc comments to structs, enums, traits and functions
- [ ] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I rebased on the latest `parachain` branch

lemunozm commented on runtime/altair/src/lib.rs:
```diff
@@ -223,8 +223,8 @@ impl Contains<Call> for BaseCallFilter {
 parameter_types! {
-	pub const ReservedXcmpWeight: Weight = MAXIMUM_BLOCK_WEIGHT / 4;
-	pub const ReservedDmpWeight: Weight = MAXIMUM_BLOCK_WEIGHT / 4;
+	pub const ReservedXcmpWeight: Weight = MAXIMUM_BLOCK_WEIGHT.saturating_div(4);
```
Is these changes done automatically by the upgrade tool? This division is checked at compile time so I think there is no need to check if saturates.

NunoAlexandre replied:
I don't know, just picked this up - cc @gruberb?

gruberb replied:
I wasn't sure about this one yet @lemunozm. The normal `div` is not `const`, so I needed another way of dividing and assigning it to a `const`. 

I am not sure, I remember @mustermeiszer opened a PR to add `const div` against the substrate repo?

lemunozm replied:
I think `div` is const if both arguments are const. I've done a fast example in [playrust](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8eb84c3b4cb2a4a4f3ab1a97b9c05af1)

gruberb replied:
We are getting this error here:

```
error[E0015]: cannot call non-const operator in constants
   --> libs/primitives/src/lib.rs:204:43
    |
204 |     pub const MAXIMUM_BLOCK_WEIGHT: Weight = WEIGHT_PER_SECOND / 2;
    |                                              ^^^^^^^^^^^^^^^^^^^^^
    |
note: impl defined here, but it is not `const`
   --> /Users/gruberbastian/.cargo/git/checkouts/substrate-7e08433d4c370a21/cc370aa/frame/support/src/weights/weight_v2.rs:236:1
    |
236 | / impl<T> Div<T> for Weight
237 | | where
238 | |     u64: Div<T, Output = u64>,
239 | |     T: Copy,
...   |
244 | |     }
245 | | }
    | |_^
    = note: calls in constants are limited to constant functions, tuple structs and tuple variants
```

lemunozm replied:
mmm... Ok, I see that `Weight` it's no longer an `u64`. And for custom types `div` is not const ([yet](https://doc.rust-lang.org/beta/unstable-book/library-features/const-ops.html)). 
Sorry for the noise and thanks for testing 🤗

NunoAlexandre replied:
Thanks for chipping in @lemunozm 🤘🏻

mustermeiszer replied:
@lemunozm @NunoAlexandre there is a PR that it already merged where we have a const div possibility. https://github.com/paritytech/substrate/pull/12322

But this will only hit v0.9.30

lemunozm replied:
Love it!

NunoAlexandre commented on pallets/collator-allowlist/Cargo.toml:
```diff
@@ -13,18 +13,18 @@ targets = ['x86_64-unknown-linux-gnu']
-sp-core = { git = "https://github.com/paritytech/substrate", default-features = true , branch = "polkadot-v0.9.26" }
-sp-io = { git = "https://github.com/paritytech/substrate", default-features = true , branch = "polkadot-v0.9.26" }
-pallet-balances = { git = "https://github.com/paritytech/substrate",  default-features = true , branch = "polkadot-v0.9.26" }
+sp-core = { git = "https://github.com/paritytech/substrate", default-features = false , branch = "polkadot-v0.9.29" }
```
The `dev-dependencies` can have the default-features enabled since they only work in std, not in wasm.

NunoAlexandre commented on pallets/crowdloan-claim/Cargo.toml:
```diff
@@ -18,26 +18,32 @@ serde = { version = "1.0.119" }
+frame-benchmarking = { git = "https://github.com/paritytech/substrate",  default-features = false , optional = true , branch = "polkadot-v0.9.29" }
+schnorrkel = { version = "0.10.2", features = ["u64_backend"], default-features = false, optional = true }
+getrandom = { version = "0.2", default-features = false, optional = true }
+rand_core = { version = "0.6.2", default-features = false, optional = true}
```
And I just realized this error I described in that commit is the one I reported in my comment of yesterday. We have officially have come full circle 🙃

NunoAlexandre commented:
Got it! 

So get around the `getrandom` issue, I've added it as an explicit dependency with `default-features = false` and `features = ["js"]' to the root Cargo, and to each of the runtime's Cargo files. I may be wrong but when I just tried adding it to the root Cargo it didn't' seem to work but we should double check.

Past that point I bumped into the `secp256k1-sys`error reported [here](https://github.com/centrifuge/centrifuge-chain/pull/1041#issuecomment-1282402385); On my machine, an Apple M1, I was able to get around that by installing `zig` (`brew install zig`) and retried building with `CC="zig cc" cargo build --release`. I then faced some substrate-relaed compilation errors, symtoms of some `default-features` not being set properly (it failed to compile some substrate crates because it didn't know `vec![]` etc).

At this point I built each runtime separately and realised that `altair` compiled fine but `centrifuge` and `development` failed. So I cherry picked which crates those two runtimes depended on that altair did not and found out that our `pallet-claims` had some substrate dependencies with `default-features = true`; setting those to `false` did the trick, and now we are back to getting the normal and expected update compilation errors 🎉  

cc @gruberb

NunoAlexandre commented:
@gruberb I've just cleaned up the changes I added for the `getrandom` fix. Let me know if I can help any further, otherwise leaving the rest up to you 👍

gruberb commented on runtime/centrifuge/src/lib.rs:
```diff
@@ -1252,8 +1252,8 @@ impl_runtime_apis! {
 		}
-		fn execute_block_no_check(block: Block) -> Weight {
-			Executive::execute_block_no_check(block)
+		fn execute_block(block: Block, state_root_check: bool, try_state: TryStateSelect) -> Weight;
```
TODO: Have to figure out how this is actually used vs. and if we used `execute_block_no_check` in the past.

mustermeiszer replied:
ASAIR the no-check prevents digest checks for block building. Do not think we can anything about the change and hopefully parity has this correctly ^^

gruberb commented:
This PR is finally ready for review! I am currently trying to figure out why one `loans` benchmark test is failing, but the logic etc. can be reviewed!

mustermeiszer commented on pallets/nft/Cargo.toml:
```diff
@@ -15,33 +15,34 @@ targets = ['x86_64-unknown-linux-gnu']
-pallet-authorship = { git = "https://github.com/paritytech/substrate", default-features = true, branch = "polkadot-v0.9.26" }
-pallet-timestamp = { git = "https://github.com/paritytech/substrate", default-features = true, branch = "polkadot-v0.9.26" }
-proofs = { path = "../../libs/proofs", default-features = true }
+pallet-anchors = { path = "../anchors", default-features = false }
+pallet-authorship = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.29" }
+pallet-timestamp = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.29" }
+proofs = { path = "../../libs/proofs", default-features = false }
```
defaults true

mustermeiszer commented on pallets/nft/Cargo.toml:
```diff
@@ -15,33 +15,34 @@ targets = ['x86_64-unknown-linux-gnu']
-sp-runtime = { git = "https://github.com/paritytech/substrate", default-features = true, branch = "polkadot-v0.9.26" }
-sp-std = { git = "https://github.com/paritytech/substrate", default-features = true, branch = "polkadot-v0.9.26" }
-cfg-types = { path = "../../libs/types", default-features = true}
+sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.29" }
+sp-runtime = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.29" }
+sp-std = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.29" }
```
defaults true

mustermeiszer commented on runtime/integration-tests/Cargo.toml:
```diff
@@ -9,88 +9,89 @@ repository = "https://github.com/centrifuge/centrifuge-chain"
+fudge = { git = "https://github.com/centrifuge/fudge", branch = "polkadot-v0.9.29" }
 tokio = { version = "1.15", features = ["macros"] }
-tracing = "0.1"
+tracing = "0.1.36"
```
Why did we need to bump this? Compatability with Fudge or Substrate?

gruberb replied:
Reverted! Must have slipped in during updating dependencies!

mustermeiszer commented on runtime/integration-tests/Cargo.toml:
```diff
@@ -9,88 +9,89 @@ repository = "https://github.com/centrifuge/centrifuge-chain"
 pallet-permissions = { path = "../../pallets/permissions" }
+pallet-pools = { path = "../../pallets/pools" }
+sp-io = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.29" }
+sp-std = { git = "https://github.com/paritytech/substrate",  default-features = false , branch = "polkadot-v0.9.29" }
+xcm-executor = { git = "https://github.com/paritytech/polkadot", default-features = false , branch = "release-v0.9.29" }
```
defaults true

mustermeiszer commented on runtime/integration-tests/src/pools/utils/env.rs:
```diff
@@ -704,13 +711,14 @@ fn test_env(
+		// let instance = INSTANCE_COUNTER.fetch_add(1, sp_std::sync::atomic::Ordering::SeqCst);
+		let instance_id = FudgeInherentTimestamp::create_instance(Default::default(), None);
+		assert!(FudgeInherentTimestamp::force_new(
+			instance_id,
 			std::time::Duration::from_secs(6),
 			Some(std::time::Duration::from_millis(START_DATE)),
 		)
-		.is_none());
+		.is_ok());
```
Remove and adapt `create_instance` to:

```rust
let instance_id = FudgeInherentTimestamp::create_instance(std::time::Duration::from_secs(6), Some(std::time::Duration::from_millis(START_DATE));
```

mustermeiszer commented on runtime/integration-tests/src/pools/utils/env.rs:
```diff
@@ -784,18 +792,19 @@ fn test_env(
-		assert!(FudgeInherentTimestamp::new(
-			instance,
+		let instance_id =
+			FudgeInherentTimestamp::create_instance(sp_std::time::Duration::from_secs(1), None);
+		assert!(FudgeInherentTimestamp::force_new(
+			instance_id,
 			std::time::Duration::from_secs(12),
 			Some(std::time::Duration::from_millis(START_DATE)),
 		)
-		.is_none());
+		.is_ok());
```
Remove and adapt create_instance to:

```rust
let instance_id = FudgeInherentTimestamp::create_instance(std::time::Duration::from_secs(12), Some(std::time::Duration::from_millis(START_DATE));
```

mustermeiszer commented on scripts/frame-weight-template.hbs:
```diff
@@ -35,22 +35,22 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 			// Standard Error: {{underscore cw.error}}
-			.saturating_add(({{underscore cw.slope}} as Weight).saturating_mul({{cw.name}} as Weight))
+			.saturating_add(Weight::from_ref_time({{underscore cw.slope}}).saturating_mul({{cw.name}}_u64))
 			{{~/each}}
 			{{~#if (ne benchmark.base_reads "0")}}
-			.saturating_add(T::DbWeight::get().reads({{benchmark.base_reads}} as Weight))
+			.saturating_add(T::DbWeight::get().reads({{benchmark.base_reads}}_u64))
 			{{~/if}}
 			{{~#each benchmark.component_reads as |cr|}}
-			.saturating_add(T::DbWeight::get().reads(({{cr.slope}} as Weight).saturating_mul({{cr.name}} as Weight)))
+			.saturating_add(T::DbWeight::get().reads(({{cr.slope}}_u64).saturating_mul({{cr.name}}_u64)))
 			{{~/each}}
 			{{~#if (ne benchmark.base_writes "0")}}
-			.saturating_add(T::DbWeight::get().writes({{benchmark.base_writes}} as Weight))
+			.saturating_add(T::DbWeight::get().writes({{benchmark.base_writes}}_u64))
 			{{~/if}}
 			{{~#each benchmark.component_writes as |cw|}}
-			.saturating_add(T::DbWeight::get().writes(({{cw.slope}} as Weight).saturating_mul({{cw.name}} as Weight)))
+			.saturating_add(T::DbWeight::get().writes(({{cw.slope}}_u64).saturating_mul({{cw.name}}_u64)))
```
@gruberb did you test creating the benchmark files with the adapted templates?

gruberb replied:
Local benchmark tests ran (finally) through. I updated the script. `clippy` tells us we have to use `5_u64` but we can't use this notation with variables (like `p_u64`). Therefore I adapted the script and used `as u64` everywhere.

mustermeiszer commented on scripts/runtime-weight-template.hbs:
```diff
@@ -25,22 +25,22 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 			// Standard Error: {{underscore cw.error}}
-			.saturating_add(({{underscore cw.slope}} as Weight).saturating_mul({{cw.name}} as Weight))
+			.saturating_add(Weight::from_ref_time({{underscore cw.slope}}).saturating_mul({{cw.name}}_u64))
 			{{~/each}}
 			{{~#if (ne benchmark.base_reads "0")}}
-			.saturating_add(T::DbWeight::get().reads({{benchmark.base_reads}} as Weight))
+			.saturating_add(T::DbWeight::get().reads({{benchmark.base_reads}}_u64))
 			{{~/if}}
 			{{~#each benchmark.component_reads as |cr|}}
-			.saturating_add(T::DbWeight::get().reads(({{cr.slope}} as Weight).saturating_mul({{cr.name}} as Weight)))
+			.saturating_add(T::DbWeight::get().reads(({{cr.slope}}_u64).saturating_mul({{cr.name}}_u64)))
 			{{~/each}}
 			{{~#if (ne benchmark.base_writes "0")}}
-			.saturating_add(T::DbWeight::get().writes({{benchmark.base_writes}} as Weight))
+			.saturating_add(T::DbWeight::get().writes({{benchmark.base_writes}}_u64))
 			{{~/if}}
 			{{~#each benchmark.component_writes as |cw|}}
-			.saturating_add(T::DbWeight::get().writes(({{cw.slope}} as Weight).saturating_mul({{cw.name}} as Weight)))
+			.saturating_add(T::DbWeight::get().writes(({{cw.slope}}_u64).saturating_mul({{cw.name}}_u64)))
```
@gruberb same question as above.

mustermeiszer commented on src/service.rs:
```diff
@@ -298,10 +298,6 @@ where
 		bool,
 	) -> Result<Box<dyn ParachainConsensus<Block>>, sc_service::Error>,
 {
-	if matches!(parachain_config.role, Role::Light) {
-		return Err("Light client not supported!".into());
-	}
-
```
Is the `Role` enum removed or why do we not check this anymore?

gruberb replied:
`Role` got removed, yes...

gruberb commented:
All green now, and comments from @mustermeiszer resolved! Should we try to merge this soon? @branan @NunoAlexandre

mustermeiszer commented on runtime/integration-tests/src/pools/utils/env.rs:
```diff
@@ -704,13 +711,17 @@ fn test_env(
+			Some(std::time::Duration::from_millis(START_DATE)),
+		);
+		assert!(FudgeInherentTimestamp::force_new(
+			instance_id,
 			std::time::Duration::from_secs(6),
 			Some(std::time::Duration::from_millis(START_DATE)),
 		)
-		.is_none());
+		.is_ok());
```
Remove. Instance is already created above.

gruberb replied:
Done.

mustermeiszer commented on runtime/integration-tests/src/pools/utils/env.rs:
```diff
@@ -784,18 +795,21 @@ fn test_env(
+			Some(std::time::Duration::from_millis(START_DATE)),
+		);
+		assert!(FudgeInherentTimestamp::force_new(
+			instance_id,
 			std::time::Duration::from_secs(12),
 			Some(std::time::Duration::from_millis(START_DATE)),
 		)
-		.is_none());
+		.is_ok());
```
Remove. Instance is already created above.

mustermeiszer reviewed (changes requested):
@gruberb  Just a small change needed. Then I will make this green ^^
